ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java

```diff
@@ -105,7 +101,8 @@ private HeartbeatAgent(final KsqlEngine engine,
     this.serviceManager = new ServiceManager(Arrays.asList(
         new DiscoverClusterService(), new SendHeartbeatService(), new CheckHeartbeatService()));
     this.receivedHeartbeats = new ConcurrentHashMap<>();
-    this.hostsStatus = new ConcurrentHashMap<>();
+    this.immutableHostsStatus = new AtomicReference<>();
```

Comment:
I would be more comfortable if we split this into two PRs: the first makes the concurrency model clearer (avoiding concurrent maps, doing copy-on-write) and the second implementing the pull query routing. that way we can focus on each individual problem